### PR TITLE
Automatically applying Scrollbars on desktop platforms with configurable ScrollBehaviors

### LINKED ISSUE #40107: Scrollbars should be always visible and instantiated by default on web and desktop
As mentioned in #38795, the the browser's scrollbar isn't currently visible when content can be scrolled. Ideally, we'd show the scrollbar when content can be scrolled otherwise this breaks users expectations on the platform.

### LINKED ISSUE #70866: ☂️ Scrollbar Fixes & Updates
This is an umbrella issue that covers several related issues for work being done on scrollbars.

- [x] Refactor Scrollbar and CupertinoScrollbar to share common methods
  - A couple of different RawScrollbar options have been discussed (https://github.com/flutter/flutter/pull/70850, https://github.com/flutter/flutter/pull/70798), the later being the current agreed upon direction.
  - [x] We should also fix this while we're at it: https://github.com/flutter/flutter/issues/70105
    - This is caused by explicitly creating a drag gesture in CupertinoScrollbar when dragging on the thumb. Instead, we should just update the position. Fix already produced in #70798
- [x] https://github.com/flutter/flutter/issues/31753: Scrollbars should be fully interactive
  - This includes dragging the thumb, tapping the track
  - CupertinoScrollbar already supports thumb dragging
- [ ] https://github.com/flutter/flutter/issues/40107: Scrollbars should be created by default for scrollables on Web and Desktop
  - will need to find a way to identify and exclude some cases, like TextFields and PageViews
- [x] https://github.com/flutter/flutter/issues/61362: Scrollbars should have a theme
  - Allows for customization across the app

Initial design doc:
flutter.dev/go/update-scrollbars

### PR DESCRIPTION
**Pull request by Piinks**
Originally https://github.com/flutter/flutter/pull/75354

[Design Document](https://docs.google.com/document/d/1W5PSIO1cVj7v6_xayydyFN3abWFOJTmFHKGBA9VKbbg/edit?usp=sharing&resourcekey=0-AHb9dJP8Knv3GWwZLnzZWQ)

This change applies `Raw`/`Cupertino`/`Scrollbar`s by default on the Web and Desktop platforms. I've done this in a similar way to how we apply the `GlowingOverscrollIndicator`, in each `Widget`/`Material`/`Cupertino` `ScrollBehavior`. This allows us to get the right `Scrollbar` for the given `ScrollBehavior`.

---

~This current proposal requires a breaking change to `ScrollBehavior.buildViewportChrome`, adding an optional `controller` parameter as `Scrollbar` requires a `ScrollController` for full functionality.~ Update: Instead I have deprecated `buildViewportChrome` in favor of `buildViewportDecoration`. Breaking the former was a really hard break, so I have introduced a new method, and added an opt-in flag (`ScrollBehavior.useDecoration`) so customers can be migrated gracefully.

This new method will use (also new) `ScrollableDetails` to ascertain the right decoration. Passing `ScrollableDetails` will be a bit more future-proof, where if needed we can add to the details rather than needing to change a function signature.  The ScrollableDetails currently contains the required AxisDirection (used for the GlowingOverscrollIndicator decoration), and an optional controller. If a controller is not provided, no `Scrollbar` will be created.

---

Some instances of `Scrollable` are excluded from this including:
- `EditableText`
- `ListWheelScrollView`
- `PageView`
- `NestedScrollView`
- Dropdown menus (they have their own `ScrollBehavior` that I have, for now, excluded scrollbars from)

~To give users control over this, I have added the ability to toggle this feature on and off in `ThemeData` and `CupertinoThemeData`.~ Update: This was a bit over the top, exposing ScrollBehaviors is a cleaner way for users to control defaults. That will be a prerequisite change, in https://github.com/flutter/flutter/pull/76739

---

~Blocked by: https://github.com/flutter/flutter/pull/76739~ 
 
Fixes: https://github.com/flutter/flutter/issues/40107
Fixes: https://github.com/flutter/flutter/issues/70866 (Last part)
Also related as part of ScrollBehavior overhaul: https://github.com/flutter/flutter/pull/75728

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/widgets/scroll_configuration.dart [hunk 1]
```diff
@@ -32,7 +35,52 @@ const Color _kDefaultGlowColor = Color(0xFFFFFFFF);
...
+  // in wrapping the Scrollable widget.
+  //
+  // This is used to maintain subclass behavior to allow for graceful migration.
+  final bool _useDecoration;
+
+  ///
+  final Set<TargetPlatform>? scrollbarPlatforms;
+
+  ///
+  static Set<TargetPlatform> get defaultScrollbarPlatforms => <TargetPlatform>{
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/scroll_configuration.dart:70]: One problem with these sets is: If we add a new TargetPlatform we will not be prompted by the analyzer to consider whether it is a scrollPlatform of glowingPlatforms...

File: packages/flutter/lib/src/cupertino/app.dart [hunk 2]
```diff
@@ -427,9 +429,27 @@ class CupertinoApp extends StatefulWidget {
...
+
   @override
-  Widget buildViewportChrome(BuildContext context, Widget child, AxisDirection axisDirection) {
-    // Never build any overscroll glow indicators.
+  Widget buildViewportDecoration(
+    BuildContext context,
+    Widget child,
+    ScrollableDetails details,
+  ) {
+    if ((scrollbarPlatforms ?? ScrollBehavior.defaultScrollbarPlatforms).contains(getPlatform(context))) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/cupertino/app.dart:447]: Could `ScrollBehavior.defaultScrollbarPlatforms` just be the default value for `scrollbarPlatforms`?

File: packages/flutter/lib/src/material/dropdown.dart [hunk 3]
```diff
@@ -90,14 +91,9 @@ class _DropdownMenuPainter extends CustomPainter {
...
-class _DropdownScrollBehavior extends ScrollBehavior {
-  const _DropdownScrollBehavior();
-
-  @override
-  TargetPlatform getPlatform(BuildContext context) => Theme.of(context).platform;
-
-  @override
-  Widget buildViewportChrome(BuildContext context, Widget child, AxisDirection axisDirection) => child;
+// The default scrollbar platforms will apply.
+class _DropdownScrollBehavior extends MaterialScrollBehavior {
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/dropdown.dart:95]: If you have a custom ScrollConfiguration higher in the tree that builds custom non-material scrollbars, this widget would have unmatching material scrollbars...

Comment (goderbauer) [reply packages/flutter/lib/src/material/dropdown.dart:95]: Presumably, we want to use the copy-with approach here as well...

File: packages/flutter/lib/src/material/app.dart [hunk 4]
```diff
@@ -688,29 +689,51 @@ class MaterialApp extends StatefulWidget {
 ///
 ///  * [ScrollBehavior], the default scrolling behavior extended by this class.
 class MaterialScrollBehavior extends ScrollBehavior {
+  /// Creates a MaterialScrollBehavior that decorates [Scrollable]s with
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/app.dart:692]: this one has to override `copyWith`?

File: packages/flutter/lib/src/cupertino/app.dart [hunk 5]
```diff
@@ -427,9 +429,27 @@ class CupertinoApp extends StatefulWidget {
 ///
 ///  * [ScrollBehavior], the default scrolling behavior extended by this class.
 class CupertinoScrollBehavior extends ScrollBehavior {
+  /// Creates a CupertinoScrollBehavior that uses [BouncingScrollPhysics] and
```
Comment (goderbauer) [inline packages/flutter/lib/src/cupertino/app.dart:432]: What about copyWith?

File: packages/flutter/lib/src/widgets/scroll_view.dart [hunk 6]
```diff
@@ -599,11 +605,74 @@ abstract class ScrollView extends StatelessWidget {
 ///    the scroll position without using a [ScrollController].
 ///  * [IndexedSemantics], which allows annotating child lists with an index
 ///    for scroll announcements.
-class CustomScrollView extends ScrollView {
+///
+class CustomScrollView extends StatelessWidget {
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/scroll_view.dart:609]: Why this change?

File: packages/flutter/lib/src/widgets/scroll_view.dart [hunk 7]
```diff
@@ -599,11 +605,74 @@ abstract class ScrollView extends StatelessWidget {
...
+      shrinkWrap: shrinkWrap,
+      center: center,
+      anchor: anchor,
+      cacheExtent: cacheExtent,
+      semanticChildCount: semanticChildCount,
+      dragStartBehavior: dragStartBehavior,
+      keyboardDismissBehavior: keyboardDismissBehavior,
+      restorationId: restorationId,
+      clipBehavior: clipBehavior,
+      scrollBehavior: scrollBehavior ?? ScrollConfiguration.of(context).copyWith(scrollbarPlatforms: <TargetPlatform>{}),
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/scroll_view.dart:670]: Why does this one not get scrollbars?

File: packages/flutter/lib/src/widgets/scrollable.dart [hunk 8]
```diff
@@ -717,7 +724,14 @@ class ScrollableState extends State<Scrollable> with TickerProviderStateMixin, R
       );
     }
 
-    return _configuration.buildViewportChrome(context, result, widget.axisDirection);
+    return _configuration.buildViewportDecoration(
+      context,
+      result,
+      ScrollableDetails(
+        direction: widget.axisDirection,
+        controller: widget.controller ?? ScrollController(),
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/scrollable.dart:732]: This scroll controller needs to be properly owned by the State: You don't want to create a new one on every build and at the end of its useful life it needs to be disposed.

File: packages/flutter/lib/src/widgets/scroll_configuration.dart [hunk 9]
```diff
@@ -32,7 +35,52 @@ const Color _kDefaultGlowColor = Color(0xFFFFFFFF);
...
   /// Creates a description of how [Scrollable] widgets should behave.
-  const ScrollBehavior();
+  const ScrollBehavior({
+    @Deprecated(
+      'Temporary migration flag, do not use. '
+      'This feature was deprecated after v2.1.0-11.0.pre.'
+    )
+    bool useDecoration = false,
+    this.scrollbarPlatforms,
+    this.glowingPlatforms,
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/scroll_configuration.dart:45]: instead of using `glowing`, maybe this should be called something a little more generic, like overscrollIndicator?

File: packages/flutter/lib/src/widgets/scroll_configuration.dart (hunk 1 above)
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/scroll_configuration.dart:70]: Instead of creating a new set every time this is called, can this just be a static const set?

**Review by goderbauer — COMMENTED** [review]
I am not convinced by the Set of TargetPlatforms:

- We loose the ability to have the analyzer remind us to consider the scrolling behavior for newly added platforms.
- We only ever seem to switch between "default platforms" and "no platforms" - why do we need the finer control?

File: packages/flutter/lib/src/widgets/scroll_view.dart (hunk 6 above)
Comment (Piinks) [reply packages/flutter/lib/src/widgets/scroll_view.dart:609]: CustomScrollView should not get a Scrollbar automatically, but with this approach of copying ScrollBehaviors to modify for individual widgets, we need to be able to get the current ScrollBehavior in order to copy it.

File: packages/flutter/lib/src/widgets/scroll_view.dart (hunk 7 above)
Comment (Piinks) [reply packages/flutter/lib/src/widgets/scroll_view.dart:670]: CustomScrollView shouldn't get scrollbars because the Scrollbar cannot automatically handle whatever folks are building in the CustomScrollView, like avoiding overlapping SliverAppBars for example.

**Comment by Piinks** [comment]
> I am not convinced by the Set of TargetPlatforms:
>
> * We loose the ability to have the analyzer remind us to consider the scrolling behavior for newly added platforms.

This is certainly a drawback. There are pros and cons to each implementation. If this makes you prefer #75354, we can move forward with that one instead. I think either or will work.

> * We only ever seem to switch between "default platforms" and "no platforms" - why do we need the finer control?

I'm trying to think ahead, right now the use cases are simpler, but future changes will introduce more complexity. So I am hoping to define a pattern that is repeatable and affords us some flexibility to build more on ScrollBehaviors in the future without continuing to break them. 

Like with the disabling drag feature on desktop, I have a rough idea that is similar where we have a Map of each PointerDeviceKind to a Set<TargetPlatform>.

Also, while our current platform configurations are simple (in the framework), users may want this granularity in their own apps to control platform specific behaviors. There have been a lot of requests for this.

**Comment by Piinks** [comment]
This needs a bunch of updates, I have notes from a brainstorm session session. I think we have an implementation we all agree on now, I just need to get typing. :)

File: packages/flutter/lib/src/widgets/scroll_configuration.dart [hunk 10]
```diff
@@ -129,6 +203,71 @@ class ScrollBehavior {
   String toString() => objectRuntimeType(this, 'ScrollBehavior');
 }
 
+class _WrappedScrollBehavior implements ScrollBehavior {
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/scroll_configuration.dart:206]: @goderbauer this does not appear to work in the way you thought it might. When `_WrappedScrollBehavior` calls `delegate.buildViewportDecoration`, it does not in turn call `_WrappedScrollBehavior.buildScrollbar`, it calls `ScrollBehavior.buildScrollbar`. It ignores the implemented class once you call on the delegate. Maybe I misunderstood what you preferred?

Comment (goderbauer) [reply packages/flutter/lib/src/widgets/scroll_configuration.dart:206]: Right, I think _WrappedScrollBehavior will have to re-implement buildViewportDecoration and call those methods on itself.

Comment (Piinks) [reply packages/flutter/lib/src/widgets/scroll_configuration.dart:206]: Won't that create an issue though if the `ScrollBehavior` being copied to `_WrappedScrollBehavior` has a different implementation of `buildViewportDecoration`? 
This is why I was trying to keep everything contained in `buildViewportDecoration` previously.

File: packages/flutter/lib/src/material/dropdown.dart [hunk 11]
```diff
@@ -90,14 +90,17 @@ class _DropdownMenuPainter extends CustomPainter {
 
 // Do not use the platform-specific default scroll configuration.
 // Dropdown menus should never overscroll or display an overscroll indicator.
+// The default scrollbar platforms will apply.
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/dropdown.dart:93]: It will use the scrollbars defined in ScrollBehavior, but it should be using the ones from the behavior in the context, right?

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 12]
```diff
@@ -1304,6 +1306,23 @@ class EditableText extends StatefulWidget {
...
+  /// {@template flutter.widgets.shadow.scrollBehavior}
+  /// A [ScrollBehavior] that will be applied to this widget individually.
+  ///
+  /// Defaults to null, wherein the inherited [ScrollBehavior] is copied and
+  /// modified to alter the viewport decoration, like [Scrollbar]s.
+  /// {@endtemplate}
+  ///
+  /// [ScrollBehavior]s also provide [ScrollPhysics]. If an explicit
+  /// [ScrollPhysics] is provided in [scrollPhysics], it will take precedence,
+  /// followed by [scrollBehavior], and then the inherited ancestor
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/editable_text.dart:1318]: We should also document this on `scrollPhysics`.

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 13]
```diff
@@ -2591,9 +2610,10 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
         excludeFromSemantics: true,
         axisDirection: _isMultiline ? AxisDirection.down : AxisDirection.right,
         controller: _scrollController,
-        physics: widget.scrollPhysics,
+        physics: widget.scrollPhysics ?? widget.scrollBehavior?.getScrollPhysics(context),
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/editable_text.dart:2613]: Is this actually necessary? When `widget.scrollPhysics` is null, it should be the `Scrollable` that does this lookup?

File: packages/flutter/lib/src/widgets/list_wheel_scroll_view.dart [hunk 14]
```diff
@@ -716,6 +721,17 @@ class ListWheelScrollView extends StatefulWidget {
   /// {@macro flutter.widgets.scrollable.restorationId}
   final String? restorationId;
 
+  /// {@macro flutter.widgets.shadow.scrollBehavior}
+  ///
+  /// [ScrollBehavior]s also provide [ScrollPhysics]. If an explicit
+  /// [ScrollPhysics] is provided in [physics], it will take precedence,
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/list_wheel_scroll_view.dart:727]: document this on physics as well?

File: packages/flutter/lib/src/widgets/list_wheel_scroll_view.dart [hunk 15]
```diff
@@ -766,9 +782,10 @@ class _ListWheelScrollViewState extends State<ListWheelScrollView> {
       },
       child: _FixedExtentScrollable(
         controller: scrollController,
-        physics: widget.physics,
+        physics: widget.physics ?? widget.scrollBehavior?.getScrollPhysics(context),
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/list_wheel_scroll_view.dart:785]: same as above

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 16]
```diff
@@ -453,6 +455,20 @@ class NestedScrollView extends StatefulWidget {
   /// {@macro flutter.widgets.scrollable.restorationId}
   final String? restorationId;
 
+  /// {@macro flutter.widgets.shadow.scrollBehavior}
+  ///
+  /// [ScrollBehavior]s also provide [ScrollPhysics]. If an explicit
+  /// [ScrollPhysics] is provided in [physics], it will take precedence,
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart:461]: Document this on physics as well?

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 17]
```diff
@@ -613,29 +629,34 @@ class NestedScrollViewState extends State<NestedScrollView> {
...
-              _coordinator!._innerController,
-              _lastHasScrolledBody!,
-            ),
-            handle: _absorberHandle,
-            clipBehavior: widget.clipBehavior,
-            restorationId: widget.restorationId,
-          );
-        },
+      child: ScrollConfiguration(
+        behavior: widget.scrollBehavior ?? ScrollConfiguration.of(context).copyWith(scrollbars: false),
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart:639]: This will disable scrollbars for all descendant scrollable. Is that what we want here? Or should we only disable it for the _NestedScrollViewCustomScrollView?

File: packages/flutter/lib/src/widgets/page_view.dart [hunk 18]
```diff
@@ -854,6 +858,17 @@ class PageView extends StatefulWidget {
   /// Defaults to [Clip.hardEdge].
   final Clip clipBehavior;
 
+  /// {@macro flutter.widgets.shadow.scrollBehavior}
+  ///
+  /// [ScrollBehavior]s also provide [ScrollPhysics]. If an explicit
+  /// [ScrollPhysics] is provided in [physics], it will take precedence,
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/page_view.dart:864]: document this on physics as well

File: packages/flutter/lib/src/widgets/scrollable.dart [hunk 19]
```diff
@@ -243,6 +243,14 @@ class Scrollable extends StatefulWidget {
   /// {@endtemplate}
   final String? restorationId;
 
+  /// {@macro flutter.widgets.shadow.scrollBehavior}
+  ///
+  /// [ScrollBehavior]s also provide [ScrollPhysics]. If an explicit
+  /// [ScrollPhysics] is provided in [physics], it will take precedence,
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/scrollable.dart:249]: also document this on physics?

File: packages/flutter/lib/src/widgets/scrollable.dart [hunk 20]
```diff
@@ -426,15 +437,21 @@ class ScrollableState extends State<Scrollable> with TickerProviderStateMixin, R
     ServicesBinding.instance!.restorationManager.flushData();
   }
 
+  @override
+  void initState() {
+    _scrollController = ScrollController();
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/scrollable.dart:442]: We should probably only create this one if widget.controller is null. In `didUpdateWidget` we will have to dispose/create the ScrollController if `widget.controller` changes.

File: packages/flutter/lib/src/widgets/scroll_configuration.dart [hunk 21]
```diff
@@ -32,7 +39,31 @@ const Color _kDefaultGlowColor = Color(0xFFFFFFFF);
 @immutable
 class ScrollBehavior {
   /// Creates a description of how [Scrollable] widgets should behave.
-  const ScrollBehavior();
+  const ScrollBehavior({
+    @Deprecated(
+      'Temporary migration flag, do not use. '
+      'This feature was deprecated after v2.1.0-11.0.pre.'
+    )
+    bool useDecoration = false,
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/scroll_configuration.dart:47]: Do you think we could avoid this flag altogether by making buildOverscrollIndicator just call buildViewportChrome by default?

File: packages/flutter/lib/src/widgets/scrollable.dart [hunk 22]
```diff
@@ -717,7 +744,14 @@ class ScrollableState extends State<Scrollable> with TickerProviderStateMixin, R
       );
     }
 
-    return _configuration.buildViewportChrome(context, result, widget.axisDirection);
+    return _configuration.buildViewportDecoration(
+      context,
+      result,
+      ScrollableDetails(
+        direction: widget.axisDirection,
+        controller: widget.controller ?? _scrollController,
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/scrollable.dart:752]: Might be useful to factor this out into a `_effectiveScrollController` getter.

**Review by goderbauer — APPROVED** [review]
LGTM assuming tests on PR and in google pass

🥳  Thanks for sticking with this!

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 23]
```diff
@@ -2609,6 +2632,7 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
         physics: widget.scrollPhysics,
         dragStartBehavior: widget.dragStartBehavior,
         restorationId: widget.restorationId,
+        scrollBehavior: widget.scrollBehavior ?? ScrollConfiguration.of(context).copyWith(scrollbars: _isMultiline),
```
[the remark this answers is not in the document]
Comment (Piinks) [reply packages/flutter/lib/src/widgets/editable_text.dart:2635]: I think it's the other way around, if `_isMultiLine` is true, we want the ScrollBehavior that will apply scrollbars. If it is false, then we copy to remove the scrollbar.

**Comment by Oleksa3** [comment]
@Piinks, hello. Currently vertical dragging recognizer from `ListView` `Scrollable` under `TextField` interferes with multiline text selection by vertical mouse move on desktop. Will it be fixed? Probably switching the recognizer off in `ListView` would be a nice option for desktop with mouse.
